### DISCUSSION TITLE
Can't use my own custom font

### DISCUSSION THREAD

**Discussion by 411A** [pymupdf/PyMuPDF#2289] [Looking for help] [answered]
## Describe the bug
Hi, I'm trying to add three text boxes to a file with different custom fonts, but every approach results in a fixed font which is not my custom font. I know it may be duplicated but I searched it in issues and docs but don't know how should I pass my fonts.

## To Reproduce
Download the `.zip` file and run the `PythonFile.py`.
[GitHubPyMuPDFIssue2288.zip](https://github.com/pymupdf/PyMuPDF/files/10979920/GitHubPyMuPDFIssue2288.zip)

<details>

<summary>📦 Source Code</summary>

```python
# pip install pymupdf
import fitz

# Load the PDF file
pdf_file = "Sample.pdf"
doc = fitz.open(pdf_file)

# Load the custom font
font_file1 = r"ZahraRoosta.ttf"
font_file2 = r"CascadiaCode-Regular.ttf"
custom_font = fitz.Font(
    fontname="ZahraRoosta",
    fontfile=font_file1
)

# Loop through each page of the PDF
for page in doc:

    # Get the page dimensions
    page_width = page.rect.width
    page_height = page.rect.height
    
    # Create a text box with different font settings
    text1 = "Sample Text One"
    text1_box = fitz.Rect(70, 150, page_width-50, page_height-50)
    #page.insert_font(font_file)
    page.insert_textbox(text1_box, text1, fontsize=16, fontfile=font_file1, color=(0, 0, 0))

    text2 = "متنی دیگر"
    text2_box = fitz.Rect(70, 200, page_width-50, page_height-50)
    page.insert_textbox(text2_box, text2, fontsize=16, fontfile=font_file1, color=(0, 0, 0))

    text2 = "Another Sample Text"
    text2_box = fitz.Rect(70, 300, page_width-50, page_height-50)
    page.insert_textbox(text2_box, text2, fontsize=16, fontfile=font_file2, color=(0, 0, 0))

# Save the modified PDF file
doc.save("NewOutput.pdf")
```
</details>

## Expected behavior
Display the right section of the image instead of the left.
<img src="https://user-images.githubusercontent.com/13054619/225311606-b8afc59b-60e4-4559-8649-45f20d875664.png" width="450" height="450" />

## Your configuration
 - Operating system: Windows 11 - 22H2 - build 22621.1344
 - Python version: 3.11.1
 - PyMuPDF version: 1.21.1 (pip installation)

**Comment by JorjMcKie** [comment, marked as answer]
Your use of `page.insert_textbox()` is incorrect - see corrected version below.
Please also note, that this set of text insertions (insert_text / insert_textbox) doen't support writing right-to-left. You must either "emulate" this in your code by reverting such strings, or use `TextWriter` output, which has a right-to-left feature.

Other than that, text comes out correctly:
![grafik](https://user-images.githubusercontent.com/8290722/225321790-26e0cba0-05ad-464f-803b-a653ec831e78.png)
```python
# pip install pymupdf
import fitz

# Load the PDF file
pdf_file = "Sample.pdf"
doc = fitz.open(pdf_file)

# Load the custom font
font_file1 = r"ZahraRoosta.ttf"
font_file2 = r"CascadiaCode-Regular.ttf"

# Loop through each page of the PDF
for page in doc:
    page.insert_font(fontfile=font_file1, fontname="F0")
    page.insert_font(fontfile=font_file2, fontname="F1")
    # Get the page dimensions
    page_width = page.rect.width
    page_height = page.rect.height

    # Create a text box with different font settings
    text1 = "Sample Text One"
    text1_box = fitz.Rect(70, 150, page_width - 50, page_height - 50)
    # page.insert_font(font_file)
    page.insert_textbox(text1_box, text1, fontsize=16, fontname="F0")

    text2 = "متنی دیگر"
    text2_box = fitz.Rect(70, 200, page_width - 50, page_height - 50)
    page.insert_textbox(text2_box, text2, fontsize=16, fontname="F0")

    text2 = "Another Sample Text"
    text2_box = fitz.Rect(70, 300, page_width - 50, page_height - 50)
    page.insert_textbox(text2_box, text2, fontsize=16, fontname="F1")

# Save the modified PDF file
doc.save(__file__.replace(".py", ".pdf"))
```

**Reply by JorjMcKie** [reply]
BTW I recommend using `doc.subset_fonts()` in your situation - this will greatly reduce the file size. So instead of the last line use these two:
```python
doc.subset_fonts()
doc.ez_save(__file__.replace(".py", ".pdf"))
```

**Reply by 411A** [reply]
Thank you so much! I used `TextWriter` > `append` > `write_text`, it fixed the RTL problem but each letter is still shown separately:
![image](https://user-images.githubusercontent.com/13054619/225334388-380b44f9-4266-4f5d-95d5-045f9c1e94a3.png)
Is there any workaround to deal with this?

**Reply by JorjMcKie** [reply]
>  Is there any workaround to deal with this?

Sorry no can do 😒

**Reply by 411A** [reply]
Ok thanks, I really like the library, tried many but PyMuPDF has a good speed, though I can't use it for Persian or Arabic text.
I think one problem maybe because of [this line](https://github.com/pymupdf/PyMuPDF/blob/e4df1422eff0f05824d1a068f9880bf8d5c1ef95/fitz/utils.py#L4394) which reverses the Persian text with a zero-width joiner. Though this method is used many times & I wasn't able to fix it by myself, and there are many details to consider like the `U+200C` unicode which we use to create a non-joined word:
```
✅ بی‌فاصله
❌ بی فاصله
```
Personally, I use unicode `U200F` to correct the direction of a non-RTL word inside an RTL sentence (or vice versa by using `U200E`).
PS: I think I should contact support@artifex.com 🤔

**Reply by JorjMcKie** [reply]
Artifex is aware of this. The solution is amazingly complex. Robin Watts from Artifex has written a lucid post here in these discussions. The things is, that in several scripts, especially Hindi, Bengali etc., there exists a rich system of "ligatures", a combination of 2 or more unicodes represented by a new glyph if occurring together in a specific sequence. An example in Latin alphabet is "&": a ligature made from the contraction of letters "e" and "t", together meaning "et" = "and" in Latin.
So outputting should generally be passed through an additional thirdparty software component "harfbuzz", which is capable of detecting these situations and taking appropriate action.

**Reply by JorjMcKie** [reply]
The best way to get in contact with our development team is using the Discord channel announced in these discussions. In parallel to #pymupdf, you will also see one for #mupdf there.

**Comment by JorjMcKie** [comment]
This is now being resolved!
Will happen in the release following the one published this week (2023-11-27).
You will then be able to do things like in the following variants of your script:
[test.zip](https://github.com/pymupdf/PyMuPDF/files/13503732/test.zip)

Sample output: 
![image](https://github.com/pymupdf/PyMuPDF/assets/8290722/4f7f6f16-c17d-4cee-8199-cbca28413666)

**Reply by JorjMcKie** [reply]
If you do not want to provide own fonts, you can simply leave them out - the system will pull in the suitable font from the Google NOTO fonts.

**Comment by dalius415** [comment]
Dear @JorjMcKie I have tried to use your code where two diff fonts are in use F0 and F1, but text with F1 does not appear in pdf.
What could be the issue here? I'm sure that with the fonts i have everything is fine, cause i have tried to mix them.

**Comment by JorjMcKie** [comment]
@dalius415 - your question cannot be answered like this, as you did not provide any data to reproduce it. You neither mention your operating system or the PyMuPDF version.

**Reply by dalius415** [reply]
Thank you for your reply.
Currently working on win10, PyMuPDF version 1.23.23.
Tried to copy/paste your provided script with no luck. F1 text not appearing.

**Reply by dalius415** [reply]
Just tried on x86_64 GNU/Linux
PyMuPDF 1.23.25: Python bindings for the MuPDF 1.23.10 library (rebased implementation).
Python 3.10 running on linux (64-bit).

Same.
Here is my own code:

    # pip install pymupdf
    import fitz

    # Load the PDF file
    pdf_file = "file.pdf"
    doc = fitz.open(pdf_file)

    # Load the custom font
    font_file1 = r"AlexBrush-Regular.ttf"
    font_file2 = r"SignPainter.ttf"

    # Loop through each page of the PDF
    for page in doc:
    page.insert_font(fontfile=font_file1, fontname="F0")
    page.insert_font(fontfile=font_file2, fontname="F1")
    # Get the page dimensions
    page_width = page.rect.width
    page_height = page.rect.height

    # Create a text box with different font settings
    text1 = input("Enter name:")
    text1_box = fitz.Rect(53, 180, page_width - 50, page_height - 50)
    page.insert_textbox(text1_box, text1, fontsize=72, fontname="F0")

    text2 = input("Enter date:")
    text2_box = fitz.Rect(105, 388, page_width - 50, page_height - 50)
    page.insert_textbox(text2_box, text2, fontsize=30, fontname="F1")

    # Save the modified PDF file
    doc.save(__file__.replace(".py", ".pdf"))

**Reply by JorjMcKie** [reply]
@dalius415 
You are misunderstanding the intention of this discussion thread!
It is about languages / fonts where conventional, character-based text output methods like `Page.insert_textbox()` and `TextWriter.fill_textbox()` **_do not work_**.

My attached zip file [test.zip](https://github.com/pymupdf/PyMuPDF/files/13503732/test.zip) introduces a new way to overcome this problem using a new method `Page.insert_htmlbox()`.

Your question is about [Page.insert_textbox()](https://pymupdf.readthedocs.io/en/latest/page.html#Page.insert_textbox) again. Please read its documentation to see why things may go wrong. You did not provide your two fonts, so there is still no way to reproduce your problem. It may be a multitude of reasons: fonts not supporting some or all characters, rectangles too small (you don't check the return code), etc.

**Reply by dalius415** [reply]
Thank you @JorjMcKie 
Sorry, i'm not so familiar with python and was trying to use your script for my internests. Anyway, your last scripts helped me out :) Something is wrong with fonts so need to work around ;) Thank you once again.

**Reply by JorjMcKie** [reply]
You are welcome.
